providers/aws-s3/src/main/java/org/jclouds/aws/s3/blobstore/internal/AWSS3PutOptionsToPutObjectOptions.java

```diff
@@ -27,16 +27,45 @@
  * for AWS S3 operations.
  */
 public class AWSS3PutOptionsToPutObjectOptions implements
-Function<PutOptions, PutObjectOptions> {
+Function<PutOptions, AWSS3PutOptionsToPutObjectOptions.ReturnValue> {
+	
+	// Oh look, it's a tuple!  Or at least it would be if, you know, Java didn't have opinions about such things.
+	public static class ReturnValue {
+		
+		private final PutObjectOptions mainOptions;
+		
+		public ReturnValue(boolean useServerSideEncryption) {
+			
+			PutObjectOptions.Builder builder = PutObjectOptions.builder();
+			if (useServerSideEncryption) { builder.serverSideEncryption(); }
+			this.mainOptions = builder.build();
+		}
+
+		public PutObjectOptions getMainOptions() {
+			return mainOptions;
+		}
+
+		/* Part options are always fixed at defaults due to OPSC-7247.  This may change if we start supporting
+		 * additional S3 options (ACLs etc.) in future ops. */
+		public PutObjectOptions getPartOptions() {
+			return PutObjectOptions.DEFAULTS;
+		}
+	}
+
+	/* We can get away with this because yes/no to server-side encryption is the only axis of variation at this point. */
```

Comment:
More like the second.  When I implemented the logic for generating main and part options it became clear that the only way they varied at the moment was through the server-side encryption header; for everything else we were just using whatever was defined in the input PutOptions object.  This comment was meant to highlight that point.

In the future we _may_ find ourselves adding new headers (or dealing with existing headers) which will also need modification when they're used in a main vs. part PutObjectOptions (or vice versa).  But for now the server-side encryption header is the only header which requires such customization, thus it's the only axis of variation.